### ISSUE TITLE
Pad_OP tensorflow pads[[0,0],[3,3],[3,3],[0,0]]  after convert [0,3,3,0,0,3,3,0] should be [0,0,3,3,0,0,3,3]?

### ISSUE THREAD

**Issue by laobadao** [onnx/tensorflow-onnx#66]
onnx Pad ：

node = onnx.helper.make_node(
    'Pad',
    inputs=['x'],
    outputs=['y'],
    mode='constant',
    value=1.2,
    pads=[0, 0, 1, 3, 0, 0, 2, 4],
)
x = np.random.randn(1, 3, 4, 5).astype(np.float32)
y = np.pad(
    x,
    pad_width=((0, 0), (0, 0), (1, 2), (3, 4)),
    mode='constant',
    constant_values=1.2,
)

expect(node, inputs=[x], outputs=[y],
       name='test_constant_pad')

tensorflow ：

pads[[0,0],[3,3],[3,3],[0,0]] 

tf-onnx pad_op：

get_tensor_value():
 [[0 0]
 [3 3]
 [3 3]
 [0 0]]
transpose():
 [[0 3 3 0]
 [0 3 3 0]]
flatten(): [0 3 3 0 0 3 3 0]

**Comment by guschmue** [comment]
In above example (the tensorflow part) you'd expect the output to have shape:
[1+0, 3+6, 4+6, 5+0] which is what I get. 
The flatten output reads as 0, 3, 3, 0 at the beginning, 0, 3, 3, 0 at the end.
I added some unit test here https://github.com/onnx/tensorflow-onnx/blob/master/tests/test_backend.py#L812 to make sure.

For the onnx example:
[0, 0, 1, 3, 0, 0, 2, 4] in tensorflow is [[0, 0], [0, 0], [1, 2], [3, 4]] and the output shape is [1, 3, 7, 12]

**Comment by laobadao** [comment]
@guschmue thanks, I see. I still have a question， the  unittest of test_pad "not supported correctly in caffe2" and when I run resnet_v1_50 in caffe2 ,the log "ONNX FATAL: [enforce fail at backend.cc:670] . Caffe2 only supports padding 2D Tensor, whereas padding is [0, 3, 3, 0, 0, 3, 3, 0, ] " means I can't fix it in there , I can only modify the code of caffe2 or onnx-caffe2?

**Comment by guschmue** [comment]
Ah, yes ... that is why I had disabled the test for caffe2. I mostly test with a internal runtime which is happy with it. Not sure why caffe2 is complaining ... the example in the onnx spec uses the same shapes and I would expect them to pass. I guess we need to take a look at caffe2 code.

**Comment by laobadao** [comment]
@guschmue 
https://github.com/pytorch/pytorch/blob/master/caffe2/onnx/backend.cc

 ```
// first two dim is for batch and channel. Note that now all the values are
  // non-negative
  if (!(pads.size() == 8 &&
        (pads.Get(0) + pads.Get(1) + pads.Get(4) + pads.Get(5) == 0))) {
    CAFFE_THROW(
        "Caffe2 only supports padding 2D Tensor, whereas padding is ", str);
  }

  // rewrite the padding info
  auto* attr = attributes.AddRewrittenAttribute(pad_name);
  attr->add_ints(pads.Get(2));
  attr->add_ints(pads.Get(3));
  attr->add_ints(pads.Get(6));
  attr->add_ints(pads.Get(7));

```
So I modify some code for caffe2 in pad_op:

```
    if not (len(paddings) == 8 and set(paddings[:2] + paddings[4:6]) == {0}):
        paddings = np.array([0, 0,paddings[1], paddings[2], 0, 0, paddings[5], paddings[6]])
```

then  I get other error:

run_onnx FAIL [enforce fail at conv_op_impl.h:30] C == filter.dim32(1) * group_. Convolution op: input channels does not match: # of input channels 9 is not equal to kernel channels * group:3*1 Error from operator: 

caffe2 PadImage OP has a attr "order "， str type "NHWC" or "NCHW" and have two method to call 
bool PadImageOp<float, CPUContext>::RunOnDeviceWithOrder**NCHW**() {.....}
bool PadImageOp<float, CPUContext>::RunOnDeviceWithOrder**NHWC**() {.....}

but onnx Pad OP don't have attr like data.format to distinguish "NHWC" or "NCHW" data type.So I think maybe this place wrong ,How do you think of this？

**Comment by laobadao** [comment]
@guschmue I solve it by this.for caffe2

```
def pad_op(ctx, node, name, args):
    # T output = Pad(T input, Tpaddings paddings, @type Tpaddings)
    # T output = Pad(T data, @STRING mode, @INTS pads, @FLOAT value)
    print("paddings before:\n", node.inputs[1].get_tensor_value())
    paddings = np.array(node.inputs[1].get_tensor_value()).transpose().flatten()

    if not (len(paddings) == 8 and set(paddings[:2] + paddings[4:6]) == {0}):
        paddings = np.array([0, 0,paddings[1], paddings[2], 0, 0, paddings[5], paddings[6]]) 

    print("paddings after:\n", paddings)
    ctx.remove_input(node, node.input[1])
    node.set_attr("pads", paddings)
    node.data_format = "NHWC"
    nodes = conv_convert_inputs(ctx, node, with_kernel=False)

    return nodes
```

**Comment by guschmue** [comment]
but that violates both tensorflow and onnx spec ... I think the purpose of tf.pad() is to pad exactly as the application says.

**Comment by laobadao** [comment]
@guschmue Yes, it is unreasonable to do so, but it can solve my problem temporarily, maybe later caffe2 can handle this part better.(*^▽^*)
